sql/src/main/java/io/crate/ingestion/IngestionService.java

```diff
@@ -0,0 +1,163 @@
+
+    private final ClusterService clusterService;
+    private final Map<String, IngestionImplementation> implementations = new HashMap<>();
+    private final Schemas schemas;
+    private IngestRulesMetaData previousIngestRulesMetaData;
+
+    @Inject
+    public IngestionService(Settings settings,
+                            Schemas schemas,
+                            ClusterService clusterService,
+                            DDLClusterStateService ddlClusterStateService) {
+        super(settings);
+        this.schemas = schemas;
+        this.clusterService = clusterService;
+        ddlClusterStateService.addModifier(new IngestionDDLClusterStateModifier());
+    }
+
+    /**
+     * Register the provided @param implementation for ingest rule changes belonging to the provided @param sourceIdent.
+     */
+    public void registerImplementation(String sourceIdent, IngestionImplementation implementation) {
+        IngestionImplementation existingImplementation = implementations.put(sourceIdent, implementation);
+        if (existingImplementation != null) {
+            throw new IllegalArgumentException("There already exists an ingestion implementation registered for " +
+                                               sourceIdent);
+        }
+
+        Map<String, Set<IngestRule>> ingestionRules = getIngestRulesOrNull(clusterService.state().metaData());
+        if (ingestionRules == null) {
+            implementation.applyRules(Collections.emptySet());
+        } else {
+            implementation.applyRules(ingestionRules.getOrDefault(sourceIdent, Collections.emptySet()));
+        }
+    }
+
+    @VisibleForTesting
+    public void removeImplementationFor(String sourceIdent) {
+        implementations.remove(sourceIdent);
+    }
+
+    @Nullable
+    private Map<String, Set<IngestRule>> getIngestRulesOrNull(MetaData metaData) {
+        IngestRulesMetaData ingestRulesMetaData =
+            (IngestRulesMetaData) metaData.customs().get(IngestRulesMetaData.TYPE);
+
+        if (ingestRulesMetaData != null) {
+            return ingestRulesMetaData.getIngestRules();
+        }
+        return null;
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.metaDataChanged() == false) {
+            return;
+        }
+
+        Map<String, Set<IngestRule>> ingestionRules = getIngestionRulesIfChangedOrNull(event.state().metaData());
+        if(ingestionRules == null) {
+            return;
+        }
+        filterOutInvalidRules(ingestionRules);
+        for (Map.Entry<String, IngestionImplementation> entry : implementations.entrySet()) {
+            String sourceIdent = entry.getKey();
+            Set<IngestRule> rulesForSource = ingestionRules.getOrDefault(sourceIdent, Collections.emptySet());
+            entry.getValue().applyRules(rulesForSource);
+        }
+    }
+
+    private void filterOutInvalidRules(Map<String, Set<IngestRule>> ingestionRules) {
```

Comment:
Would it make sense to remove rules if a table is dropped? Given that we only allow rule creation on existing tables

Reply:
We actually implemented that and removed the implementation after discussing with @dobe. We decided not to implement it as it would possibly be a lot of admin to re-add the rules and they are virtually invalid anyway so won't be taken into account by implementations.